### PR TITLE
Make JSON-LD "@context" parsing more robust (closes #16006)

### LINKED ISSUE #16006: JSON-LD with "@context": "https://schema.org"  not parsed by `_json_ld` function
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.03.26.1*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.03.26.1**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The Problem

The way the  `@context` property of a JSON-LD string is checked it too strict. Currently `@context` has to be exactly the string `http://schema.org` in order to be parsed by the `_json_ld` function in `common.py`. However, some websites have `http` changed to `https`. I also saw websites adding a `/` to the end like so: `http://schema.org/`.
This leads to the problem, that JSON-LD strings of those websites do not get parsed, which cause video extraction errors, since `_json_ld` then returns an empty dictionary. (See below for an example.)

### Suggested Fix
In order to solve the problem, I suggest to make the check for the `@context` property more resiliant. Instead of checking against a static string...

https://github.com/rg3/youtube-dl/blob/5d60b9971784289acd4325a8ed7b5afd7bea05ca/youtube_dl/extractor/common.py#L1028

we should use a regex that allows the above mentioned modification of the string:
```python
if re.match(r'^https?://schema.org/?$', e.get('@context')):
```

This small change already fixes the currently broken `gamestar.py` extractor. Additionally, no existing extractor can be affected in a negative way, e.g, no extractor will get broken through this change.

Please let me know, if I should open a pull request myself.

---
### Example
An example of an extractor that fails because of not parsed JSON-LD is `gamestar.py`. Please note, that 'KeyError' occurs only because the info_dict is empty, since the `self._json_ld` function didn't parse anything. Here, there is a `https` instead of `http` used in the `@context` property.

```
# python -m youtube_dl -v http://www.gamestar.de/videos/trailer,3/hobbit-3-die-schlacht-der-fuenf-heere,76110.html
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'http://www.gamestar.de/videos/trailer,3/hobbit-3-die-schlacht-der-fuenf-heere,76110.html']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.03.26.1
[debug] Git HEAD: 5d60b9971
[debug] Python version 3.6.4 (CPython) - Linux-4.15.7-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 3.4.2, ffprobe 3.4.2
[debug] Proxy map: {}
[GameStar] 76110: Downloading webpage
ERROR: An extractor error has occurred. (caused by KeyError('title',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/home/user/youtube-dl/youtube_dl/extractor/common.py", line 440, in extract
    ie_result = self._real_extract(url)
  File "/home/user/youtube-dl/youtube_dl/extractor/gamestar.py", line 50, in _real_extract
    info_dict['title'], ' - Game%s' % site.title())
KeyError: 'title'
Traceback (most recent call last):
  File "/home/user/youtube-dl/youtube_dl/extractor/common.py", line 440, in extract
    ie_result = self._real_extract(url)
  File "/home/user/youtube-dl/youtube_dl/extractor/gamestar.py", line 50, in _real_extract
    info_dict['title'], ' - Game%s' % site.title())
KeyError: 'title'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/youtube-dl/youtube_dl/YoutubeDL.py", line 785, in extract_info
    ie_result = ie.extract(url)
  File "/home/user/youtube-dl/youtube_dl/extractor/common.py", line 453, in extract
    raise ExtractorError('An extractor error has occurred.', cause=e)
youtube_dl.utils.ExtractorError: An extractor error has occurred. (caused by KeyError('title',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

### PR DESCRIPTION
**Pull request by aeph6Ee0**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This modification makes the JSON-LD "@context" field matching more robust by using a Regex. Without this patch, JSON-LD strings are some times not recognized which leads to failure in extractors (e.g. #16006).
Furthermore, this patch will prevent extractors from failing in the future, if the "@context" field gets slightly modified by a website.

### CODE REVIEW

File: youtube_dl/extractor/common.py [hunk 1]
```diff
@@ -1025,7 +1025,7 @@ def extract_video_object(e):
             })
 
         for e in json_ld:
-            if e.get('@context') == 'http://schema.org':
+            if re.match(r'^https?://schema.org/?$', e.get('@context')):
```
Comment (dstftw) [inline youtube_dl/extractor/common.py]: Breaks if no `@context` is present.

File: youtube_dl/extractor/common.py [hunk 2]
```diff
@@ -1025,7 +1025,7 @@ def extract_video_object(e):
             })
 
         for e in json_ld:
-            if e.get('@context') == 'http://schema.org':
+            if '@context' in e and re.match(r'^https?://schema.org/?$', e.get('@context')):
```
Comment (dstftw) [inline youtube_dl/extractor/common.py]: Breaks if `e.get('@context')` is not a string.

File: youtube_dl/extractor/common.py [hunk 3]
```diff
@@ -1025,7 +1025,7 @@ def extract_video_object(e):
             })
 
         for e in json_ld:
-            if '@context' in e and re.match(r'^https?://schema.org/?$', e.get('@context')):
+            if isinstance(e.get('@context'), str) and re.match(r'^https?://schema.org/?$', e.get('@context')):
```
Comment (dstftw) [inline youtube_dl/extractor/common.py]: `compat_str`.

**Comment by aeph6Ee0** [comment]
@dstftw: Thanks for pointing out the numerous (potential) errors one at a time. This has a good learning effect.

**Pull request merged**